File: cs/vmtest/MathTests.cs
namespace vmtest;

using vm;

[TestClass]
public class MathTests
{
    [TestMethod]
    public void TestADD()
    {
        VirtualMachine vm = new VirtualMachine();

        vm.Execute(new Bytecode[] {
            Bytecode.CONST, (Bytecode)5,
            Bytecode.CONST, (Bytecode)5,
            Bytecode.ADD,
        });

        Assert.AreEqual(1, vm.Stack.Length);
        Assert.AreEqual(10, vm.Stack[0]);
    }
    [TestMethod]
    public void TestSUB()
    {
        VirtualMachine vm = new VirtualMachine();
        vm.Execute(new Bytecode[] {
            Bytecode.CONST, (Bytecode)15,
            Bytecode.CONST, (Bytecode)5,
            Bytecode.SUB,
        });

        Assert.AreEqual(1, vm.Stack.Length);
        Assert.AreEqual(10, vm.Stack[0]);
    }
    [TestMethod]
    public void TestMUL()
    {
        VirtualMachine vm = new VirtualMachine();
        vm.Execute(new Bytecode[] {
            Bytecode.CONST, (Bytecode)5,
            Bytecode.CONST, (Bytecode)2,
            Bytecode.MUL,
        });

        Assert.AreEqual(1, vm.Stack.Length);
        Assert.AreEqual(10, vm.Stack[0]);
    }
    [TestMethod]
    public void TestDIV()
    {
        VirtualMachine vm = new VirtualMachine();
        vm.Execute(new Bytecode[] {
            Bytecode.CONST, (Bytecode)20,
            Bytecode.CONST, (Bytecode)2,
            Bytecode.DIV,
        });

        Assert.AreEqual(1, vm.Stack.Length);
        Assert.AreEqual(10, vm.Stack[0]);
    }
    [TestMethod]
    public void TestMOD()
    {
        VirtualMachine vm = new VirtualMachine();
        vm.Execute(new Bytecode[] {
            Bytecode.CONST, (Bytecode)30,
            Bytecode.CONST, (Bytecode)20,
            Bytecode.MOD,
        });

        Assert.AreEqual(1, vm.Stack.Length);
        Assert.AreEqual(10, vm.Stack[0]);
    }
    [TestMethod]
    public void TestABS()
    {
        VirtualMachine vm = new VirtualMachine();
        vm.Execute(new Bytecode[] {
            Bytecode.CONST, (Bytecode)30,
            Bytecode.ABS,
        });

        Assert.AreEqual(1, vm.Stack.Length);
        Assert.AreEqual(30, vm.Stack[0]);
    }
    [TestMethod]
    public void TestNEG()
    {
        VirtualMachine vm = new VirtualMachine();
        vm.Execute(new Bytecode[] {
            Bytecode.CONST, (Bytecode)(-10),
            Bytecode.NEG,
        });

        Assert.AreEqual(1, vm.Stack.Length);
        Assert.AreEqual(10, vm.Stack[0]);
    }
    [TestMethod]
    public void TestNEGNEG()
    {
        VirtualMachine vm = new VirtualMachine();
        vm.Execute(new Bytecode[] {
            Bytecode.CONST, (Bytecode)(10),
            Bytecode.NEG,
            Bytecode.NEG,
        });

        Assert.AreEqual(1, vm.Stack.Length);
        Assert.AreEqual(10, vm.Stack[0]);
    }
}

File: cpp/tests/math_tests.cpp
//
// Created by George Liontos on 20/4/24.
//

#include "unit_test.h"
#include "../virtual_machine.h"

UNIT_TEST(add) {
    VirtualMachine vm = {};
    Bytecode codes[] = {
            Bytecode::CONST, (Bytecode) 5,
            Bytecode::CONST, (Bytecode) 5,
            Bytecode::ADD
    };

    vm_execute(&vm, codes, ARRAY_SIZE(codes));

    ASSERT(vm.stack_pointer == 0);
    ASSERT(vm.stack[0] == 10);

    TEST_PASSED();
}

UNIT_TEST(sub) {
    VirtualMachine vm = {};
    Bytecode codes[] = {
            Bytecode::CONST, (Bytecode) 15,
            Bytecode::CONST, (Bytecode) 5,
            Bytecode::SUB
    };

    vm_execute(&vm, codes, ARRAY_SIZE(codes));

    ASSERT(vm.stack_pointer == 0);
    ASSERT(vm.stack[0] == 10);

    TEST_PASSED();
}

UNIT_TEST(mul) {
    VirtualMachine vm = {};
    Bytecode codes[] = {
            Bytecode::CONST, (Bytecode) 2,
            Bytecode::CONST, (Bytecode) 5,
            Bytecode::MUL
    };

    vm_execute(&vm, codes, ARRAY_SIZE(codes));

    ASSERT(vm.stack_pointer == 0);
    ASSERT(vm.stack[0] == 10);

    TEST_PASSED();
}

UNIT_TEST(div) {
    VirtualMachine vm = {};
    Bytecode codes[] = {
            Bytecode::CONST, (Bytecode) 50,
            Bytecode::CONST, (Bytecode) 5,
            Bytecode::DIV
    };

    vm_execute(&vm, codes, ARRAY_SIZE(codes));

    ASSERT(vm.stack_pointer == 0);
    ASSERT(vm.stack[0] == 10);

    TEST_PASSED();
}

UNIT_TEST(mod) {
    VirtualMachine vm = {};
    Bytecode codes[] = {
            Bytecode::CONST, (Bytecode) 30,
            Bytecode::CONST, (Bytecode) 20,
            Bytecode::MOD
    };

    vm_execute(&vm, codes, ARRAY_SIZE(codes));

    ASSERT(vm.stack_pointer == 0);
    ASSERT(vm.stack[0] == 10);

    TEST_PASSED();
}

UNIT_TEST(abs) {
    VirtualMachine vm = {};
    Bytecode codes[] = {
            Bytecode::CONST, (Bytecode) -10,
            Bytecode::ABS
    };

    vm_execute(&vm, codes, ARRAY_SIZE(codes));

    ASSERT(vm.stack_pointer == 0);
    ASSERT(vm.stack[0] == 10);

    TEST_PASSED();
}

UNIT_TEST(neg) {
    VirtualMachine vm = {};
    Bytecode codes[] = {
            Bytecode::CONST, (Bytecode) 10,
            Bytecode::NEG
    };

    vm_execute(&vm, codes, ARRAY_SIZE(codes));

    ASSERT(vm.stack_pointer == 0);
    ASSERT(vm.stack[0] == -10);

    Bytecode neg = Bytecode::NEG;
    vm_execute(&vm, &neg, 1);

    ASSERT(vm.stack_pointer == 0);
    ASSERT(vm.stack[0] == 10);

    TEST_PASSED();
}

TestFn tests[] = {
        test_add,
        test_sub,
        test_mul,
        test_div,
        test_mod,
        test_abs,
        test_neg,
};

int main() {
    run_tests(tests, ARRAY_SIZE(tests));
    return 0;
}

